nixos/modules/flake/source-info.nix

```diff
@@ -0,0 +1,31 @@
+{ config, lib, ... }:
+
+with lib;
+
+let sourceInfo = config.system.nixos.configuration.sourceInfo;
+
+in
+{
+  options = {
+
+    system.nixos.configuration.sourceInfo = mkOption {
+      type = types.attrsOf types.anything;
+      default = {};
+      example = "self.sourceInfo";
+      description = ''
+        The source information of a flake-based NixOS configuration.
+        If set, the attribute <literal>configurationRevision</literal>
+        will appear properly in the output of
+        <literal>nixos-version --json</literal>.
+
+        Caution: Setting this option may result in unnecessary
+        re-deployments if the flake repository changes but the
+        specific system configuration does not change.
+      '';
+    };
+  };
+
+  config = {
+    system.configurationRevision = mkIf (sourceInfo ? rev) sourceInfo.rev;
```

Comment:
Can we set it to `"dirty"` when we see that `sourceInfo` is from git but does not have `rev`?
We don't have a `type`-like attribute in there (yet?), but the presence of `submodules` should give it away.

Reply:
https://github.com/NixOS/nix/issues/6747

Reply:
The intention of setting it to `dirty`, to my understanding, is to help identify the state of the configuration better. To answer this question well, I checked the relevant code, and have two proposals:

1. We can do it for all supported VCSes, not just git. For instance, mercurial is also supported (albeit immature). This is partly due to [only git getting special handling while examining local paths](https://github.com/NixOS/nix/blob/f6531851072308139b6cf15151137a1808846ba0/src/libexpr/flake/flakeref.cc#L157), and we may improve that too to enable support for other VCSes. This also implies a standardized, fine-grained specification of source information, as will be discussed in NixOS/nix#6747.

2. To further tell differences among `dirty` configurations, we can fetch the latest update time of a repository and set something like `dirty-<timestamp>`. The latest update time can be defined as the most recent modification time of all files in the repository.

Reply:
@roberth Do we solve this later as a part of NixOS/rfcs#125?

Reply:
We can scope this discussion out regardless.